paddle/fluid/operators/optimizers/momentum_op.h

```diff
@@ -242,99 +289,141 @@ class SparseMomentumFunctor;
   const int64_t row_numel_;
   const int64_t row_height_;
-  T* p_out_;
-  T* v_out_;
+  T* param_out_;
+  T* velocity_out_;
+  RegularizationFlag regularization_flag_;
+  const T regularization_coeff_;
 
  public:
-  SparseMomentumFunctor(const T* p, const T* g, const T* v, const T* lr,
-                        const T mu, const int64_t* rows, int64_t row_numel,
-                        int64_t row_height, T* p_out, T* v_out)
-      : p_(p),
-        g_(g),
-        v_(v),
+  SparseMomentumFunctor(const T* param, const T* grad, const T* velocity,
+                        const T* lr, const T mu, const int64_t* rows,
+                        int64_t row_numel, int64_t row_height,
+                        RegularizationFlag regularization_flag,
+                        const T regularization_coeff, T* param_out,
+                        T* velocity_out)
+      : param_(param),
+        grad_(grad),
+        velocity_(velocity),
         lr_(lr),
         mu_(mu),
         rows_(rows),
         row_numel_(row_numel),
         row_height_(row_height),
-        p_out_(p_out),
-        v_out_(v_out) {}
+        param_out_(param_out),
+        velocity_out_(velocity_out),
+        regularization_flag_(regularization_flag),
+        regularization_coeff_(regularization_coeff) {}
 
   inline HOSTDEVICE void operator()(size_t i) {
     auto row_idx =
         math::BinarySearch<int64_t>(rows_, row_height_, i / row_numel_);
-    T g = row_idx >= 0 ? g_[row_idx * row_numel_ + i % row_numel_]
-                       : static_cast<T>(0);
+    T grad = row_idx >= 0 ? grad_[row_idx * row_numel_ + i % row_numel_]
+                          : static_cast<T>(0);
     // put memory access in register
-    const T p = p_[i];
+    const T param = param_[i];
     const T lr = lr_[0];
-    const T v = v_[i];
-    T v_out = v * mu_ + g;
-    T p_out = p - v_out * lr;
+    const T velocity = velocity_[i];
+
+    grad = regularization_flag_ == RegularizationFlag::kL2DECAY
+               ? grad + regularization_coeff_ * param
+               : grad;
+
+    T velocity_out = velocity * mu_ + grad;
+    T param_out = param - velocity_out * lr;
     // write reigster to memory
-    v_out_[i] = v_out;
-    p_out_[i] = p_out;
+    velocity_out_[i] = velocity_out;
+    param_out_[i] = param_out;
   }
 };
 
 template <typename DeviceContext, typename T>
 class MomentumOpKernel : public framework::OpKernel<T> {
  public:
   void Compute(const framework::ExecutionContext& ctx) const override {
+    std::string regularization_method =
+        ctx.Attr<std::string>("regularization_method");
+    if (regularization_method != "" || !regularization_method.empty()) {
+      PADDLE_ENFORCE_EQ("l2_decay", regularization_method,
+                        platform::errors::InvalidArgument(
+                            "if regularization_method is not null, "
+                            "it should be %s, but received %s",
+                            regularization_method));
```

Comment:
It seems there are two `%s`, but only one parameter given.

Reply:
thanks, done.